internal/substatoptimizer/substatoptimizer.go

```diff
@@ -389,6 +389,12 @@ func RunSubstatOptim(simopt simulator.Options, verbose bool, additionalOptions s
 
 			substatGradients[idxSubstat] = substatEvalResult.DPS.Mean - initialMean
 
+			// fixes cases in which fav holders don't get enough crit rate to reliably proc fav (an important example would be fav kazuha)
+			// might give them "too much" cr (= max out liquid cr subs) but that's probably not a big deal
+			if charWithFavonius[idxChar] && idxSubstat == 1 {
+				substatGradients[idxSubstat] += 1000
```

Comment:
@k0l11 I think I understand the approach, I believe this would simply make me the optimizer significantly more likely to allocate CR in any given optimization pass and I don't doubt 1000 is a good number in testing. I think it's a reasonable approach.

Regardless, have you considered/tested instead adjusting the CR baseline here? https://github.com/genshinsim/gcsim/blob/16a53d7a21b0685cb9b6c67f6530791ae2c187d3/internal/substatoptimizer/substatoptimizer.go#L248

I wonder if it would be more consistent to increase the baseline CR provided to fav users, and to leave the CR gradient alone. It seems difficult to asses the appropriate number to supply to the gradient to correctly value CR in a fav context, but I'm also haven't dug into the problem too much nor am I very practiced with gradient descents generally. Let me know your thoughts!

Reply:
Note, if we do adjust the baseline stats, we also ought to adjust these two lines:

https://github.com/genshinsim/gcsim/blob/16a53d7a21b0685cb9b6c67f6530791ae2c187d3/internal/substatoptimizer/substatoptimizer.go#L365

https://github.com/genshinsim/gcsim/blob/16a53d7a21b0685cb9b6c67f6530791ae2c187d3/internal/substatoptimizer/substatoptimizer.go#L531

Reply:
The problem with this approach is that the CR gradient is usually so low that it gets picked last. Increasing the CR baseline will just make it give more CR at baseline but it will not let it go above that baseline because of how the gradients work out.  

For Kazuha something like this can happen without any changes to optimizer:
- optimizer has a special case for favonius that always adds 4 liquid cr subs at baseline before optimizing crit rolls
- optimizer determines 2 liquid er rolls are the best for kazuha in that sim
- optimizer sees em gives the biggest dmg increase -> give 4 em rolls, since that's the cap for an em/em/em build in kqmc
- optimizer sees atk% gives the next biggest dmg increase -> give 10 atk% rolls
- optimizer has no more rolls left to give to cr because 2 er + 4 em + 10 atk% + 4 cr = 20 liquid subs (cap of kqmc) 

In some cases atk% is missing from the list of gradients so it can give cr subs to Kazuha, but that's not consistent.

Reply:
@k0l11 Yeah I thought about it more yesterday and even increasing the CR baseline would run into the same complication with characters that want an EM hat versus a CR hat. Bummer, but makes sense to me that biasing the gradient towards CR would solve better for the kazuha case. 

One last consideration, perhaps this behavior should be limited to optimizer runs on characters that we whitelist? Kazu, venti, sucrose, Jean (maybe every anemo lmao?). I'm ambivalent on this thought, leaning towards a no because I think we always want to bias CR for fav-equipped characters, I don't think that changes based on the most relevant substats for the character. It's another question of by how much we bias, but starting with +1000 avg dmg delta seems like a reasonable enough starting point to me.

Let me know your thoughts, approving this

Reply:
I agree with giving a CR bias to characters with Fav weapons and if issues come up later we can always reconsider the bias.